More Vue Typescript Fixes

jmchilton opened the pull request:
(Please replace this header with a description of your pull request. Please include *BOTH* what you did and why you made the changes. The "why" may simply be citing a relevant Galaxy issue.)
(If fixing a bug, please add any relevant error or traceback)
(For UI components, it is recommended to include screenshots or screencasts)

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

jmchilton committed:
Fix vue-tsc error with a new abstraction in PairCollectionCreator.

jmchilton committed:
Remove unused component with a type error.

jmchilton committed:
Suppress type error in ExportForm.test.ts.

This is ugly but it seems like what we do in FilterMenu.test.ts? Any ideas to fix both more cleanly?

jmchilton committed:
Fix filtertype in FilterMenuMultiTags.

jmchilton committed:
Fix types around FilterMenuObjectStore.

jmchilton committed:
Fix ObjectStoreSelect.

jmchilton committed:
Improve typescript awareness of types in GenericWizard.

davelopez commented on client/src/components/Collections/FixedIdentifierDatasetCollectionElementView.vue:
```diff
@@ -0,0 +1,29 @@
+import localize from "@/utils/localization";
+
+interface Props {
+    element: HDASummary | undefined;
```
```suggestion
    element?: HDASummary;
```
Subjectively prettier and more consistent :)

davelopez commented on client/src/components/Collections/FixedIdentifierDatasetCollectionElementView.vue:
```diff
@@ -0,0 +1,29 @@
+defineProps<Props>();
+
+const emit = defineEmits<{
+    (event: "element-is-discarded", element: any): void;
```
If we trust the prop type:
```suggestion
    (event: "element-is-discarded", element: HDASummary): void;
```

davelopez commented on client/src/components/Collections/FixedIdentifierDatasetCollectionElementView.vue:
```diff
@@ -0,0 +1,29 @@
+    <span>
+        <DatasetCollectionElementView
+            v-if="element"
+            :key="element?.id"
```
I think v-if will ensure this will not be undefined.
```suggestion
            :key="element.id"
```

davelopez commented on client/src/components/Common/FilterMenuObjectStore.vue:
```diff
@@ -41,14 +41,14 @@ watch(
 
-function onChange(value: string | null) {
-    localValue.value = (value || undefined) as FilterType;
+function onChange(value: string | undefined) {
```
```suggestion
function onChange(value: FilterType) {
```

davelopez reviewed (commented):
This is much needed! Really appreciated!

I missed the last meeting but maybe it was discussed how to distribute the vue-tsc issues so we don't duplicate effort. I want to help too! but not sure where to start :sweat_smile:

jmchilton committed:
Apply suggestions from code review

Co-authored-by: David López <46503462+davelopez@users.noreply.github.com>

jmchilton committed:
vue-tsc type fixes for ConfigurationTestSummary.vue

jmchilton committed:
vue-tsc type fixes for config template tests

jmchilton committed:
More Vue test typing supression.

jmchilton committed:
Correct type in DataDialog to align Data and Selection dialogs.

jmchilton committed:
type fix for DatasetPermissionForm.

jmchilton committed:
Type error supression for DatasetStorage.test.ts.

jmchilton committed:
Fix subtle typing error by converging backend types in good way.

I knew these not leveraging common field descriptions or inheritence would bite us at some point - it is really nice to catch it this way in the typescript.

jmchilton committed:
Fix typescript errors for DatasetDetails.

jmchilton committed:
Fix typescript errors with DatasetError.

jmchilton committed:
Fix datatype error in datasetinformation.vue

jmchilton committed:
Stronger typing and a fix around object store relocation.

jmchilton committed:
Type issue suppression in DatasetAttributes.test.ts.

jmchilton committed:
Typescript fix in DatasetError.vue.

jmchilton committed:
Fix some small typing errors in FormElement.vue.

davelopez commented on client/src/components/Dataset/DatasetStorage/RelocateDialog.vue:
```diff
@@ -16,6 +16,17 @@ const emit = defineEmits<{
     (e: "closeModal"): void;
 }>();
 
+function relocate(objectStoreId: string | undefined | null) {
+    if (objectStoreId) {
+        // all the target object stores will have an object store id - they were
+        // generated thusly. But the type system doesn't know this. So this function
+        // wraps the emit.
+        emit("relocate", objectStoreId);
+    } else {
+        console.error("Serious logic problem, this branch should be unreachable.");
+    }
+}
+
```
A quick search for `ConcreteObjectStoreModel` in the backend shows that we always provide an ID when returning one of these as you suggest:

https://github.com/galaxyproject/galaxy/blob/5682b6a0a229f25e228a1d1d4e30772d403bf1f5/lib/galaxy/objectstore/__init__.py#L781

So maybe we can better fix the pydantic model if that is the case:

https://github.com/galaxyproject/galaxy/blob/5682b6a0a229f25e228a1d1d4e30772d403bf1f5/lib/galaxy/objectstore/__init__.py#L1772

jmchilton replied:
I appreciate the attention to detail. I mean... I think on principle the model representation of an object store can definitely exist without an object store ID though. Like the default disk object store is a concrete object store IMO and it has no ID since it isn't in a nested object store of any kind. Maybe we don't use any of these APIs but I think they could still be modeled. I think the object store APIs have grown in a sort of organic, pragmatic way and definitely have holes for things that should be implemented but are not because they are not used in the UI - I think the models are right though. Does that make sense. E.g. https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/webapps/galaxy/api/object_store.py#L82 - this is not a generic query, it is a slice of the actual of the API we would design in the abstract but tailored to a narrow UI use case (switching and selecting selectable object stores from a particular kind of distributed object store).

So my push back would be we would need to subclass ConcreteObjectStoreModel with a subclass that has guaranteed ID - maybe NestedObjectStoreConcreteChildObjectStoreModel - but that is going to result in some of the same hackage on the frontend and is messy in its own way. It doesn't feel worth the effort to cleanup the typescript IMO.

davelopez replied:
Ahh, I see! I thought it was just a typical case of "A Pydantic model that needs adjustment after real use case" because I found many of these while porting other components to Typescript :sweat_smile: 

Thanks for taking the time to explain in detail!

davelopez commented on client/src/components/DatasetInformation/DatasetError.vue:
```diff
@@ -126,6 +126,14 @@ function onMissingJobId() {
 }
 
+const userEmail = computed<string | null>(() => {
+    if (currentUser && currentUser.value && "email" in currentUser.value) {
```
```suggestion
    if (isRegisteredUser(currentUser.value)) {
```
You will need to import isRegisteredUser from `@/api` too.

jmchilton replied:
Very nice cleanup - thanks so much!

jmchilton committed:
Cleanup suggested from @davelopez
